quinn-udp/src/windows.rs

```diff
@@ -68,38 +133,162 @@ impl UdpSocketState {
+                let rc = unsafe {
+                    WinSock::setsockopt(
+                        socket.0.as_raw_socket() as _,
+                        WinSock::IPPROTO_IPV6 as _,
+                        WinSock::IPV6_PKTINFO as _,
+                        &sock_true as *const _ as _,
+                        mem::size_of_val(&sock_true) as _,
+                    )
+                };
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+
+                let rc = unsafe {
+                    WinSock::setsockopt(
+                        socket.0.as_raw_socket() as _,
+                        WinSock::IPPROTO_IPV6 as _,
+                        WinSock::IPV6_ECN as _,
+                        &sock_true as *const _ as _,
+                        mem::size_of_val(&sock_true) as _,
+                    )
+                };
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+            }
         }
 
         let now = Instant::now();
         Ok(Self {
             last_send_error: Mutex::new(now.checked_sub(2 * IO_ERROR_LOG_INTERVAL).unwrap_or(now)),
+            wsa_recvmsg_ptr,
         })
     }
 
     pub fn send(&self, socket: UdpSockRef<'_>, transmits: &[Transmit]) -> Result<usize, io::Error> {
         let mut sent = 0;
         for transmit in transmits {
-            match socket.0.send_to(
-                &transmit.contents,
-                &socket2::SockAddr::from(transmit.destination),
-            ) {
-                Ok(_) => {
-                    sent += 1;
+            // FIXME be symmetric ? do not send ECN if we cannot receive them ?
+            if self.wsa_recvmsg_ptr.is_some() {
+                // we cannot use [`socket2::sendmsg()`] and [`socket2::MsgHdr`] as we do not have access
+                // to the inner field which holds the WSAMSG
+                let mut ctrl_buf = cmsg::Aligned([0; CMSG_LEN]);
+                let daddr = socket2::SockAddr::from(transmit.destination);
+
+                let mut data = WinSock::WSABUF {
+                    buf: transmit.contents.as_ptr() as *mut _,
+                    len: transmit.contents.len() as _,
+                };
+
+                let ctrl = WinSock::WSABUF {
+                    buf: ctrl_buf.0.as_mut_ptr(),
+                    len: ctrl_buf.0.len() as _,
+                };
+
+                let mut wsa_msg = WinSock::WSAMSG {
+                    name: daddr.as_ptr() as *mut _,
+                    namelen: daddr.len(),
+                    lpBuffers: &mut data,
+                    Control: ctrl,
+                    dwBufferCount: 1,
+                    dwFlags: 0,
+                };
+
+                // Add control messages (ECN and PKTINFO)
+                let mut encoder = unsafe { cmsg::Encoder::new(&mut wsa_msg) };
+
+                let ecn = transmit.ecn.map_or(0, |x| x as u32);
```

Comment:
Can we cite docs attesting that `u32` is the proper cmsg payload type here? That's been a common source of bugs, across multiple QUIC impls.

Reply:
i will look to find a documentation to link, if i don't find one i can link to msquic userland code.